src/vmm/src/persist.rs

```diff
@@ -32,6 +43,139 @@ pub struct MicrovmState {
     pub device_states: DeviceStates,
 }
 
+/// Errors related to saving Microvm state.
+#[derive(Debug)]
+pub enum SaveMicrovmStateError {
+    /// Failed to save vCPU state.
+    SaveVcpuState(Option<vstate::Error>),
+    /// Failed to save VM state.
+    SaveVmState(vstate::Error),
+    /// Failed to send event.
+    SignalVcpu(vstate::Error),
+}
+
+impl Display for SaveMicrovmStateError {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::SaveMicrovmStateError::*;
+        match self {
+            SaveVcpuState(err) => write!(f, "Unable to save Vcpu state. Error: {:?}", err),
+            SaveVmState(err) => write!(f, "Unable to save Vm state. Error: {:?}", err),
+            SignalVcpu(err) => write!(f, "Unable to signal Vcpu: {:?}", err),
+        }
+    }
+}
+
+/// Errors associated with creating a snapshot.
+#[derive(Debug)]
+pub enum CreateSnapshotError {
+    /// Failed to get dirty bitmap.
+    DirtyBitmap,
+    /// Failed to open memory backing file.
+    MemoryBackingFile(std::io::Error),
+    /// Failed to save MicrovmState.
+    MicrovmState(SaveMicrovmStateError),
+    /// Failed to serialize microVM state.
+    SerializeMicrovmState(snapshot::Error),
+    /// Failed to open the snapshot backing file.
+    SnapshotBackingFile(std::io::Error),
+    /// Failed to write memory to snapshot.
+    Memory(memory_dump::Error),
+}
+
+impl Display for CreateSnapshotError {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::CreateSnapshotError::*;
+        match self {
+            DirtyBitmap => write!(f, "Unable to get dirty bitmap"),
+            MicrovmState(err) => write!(f, "Unable to save microvm state: {}", err),
+            MemoryBackingFile(err) => write!(f, "Unable to open mmeory file: {:?}", err),
+            SerializeMicrovmState(err) => write!(f, "Unable to serialize MicrovmState: {:?}", err),
+            SnapshotBackingFile(err) => write!(f, "Unable to open snapshot file: {:?}", err),
+            Memory(err) => write!(f, "Unable to write memory file: {:?}", err),
+        }
+    }
+}
+
+/// Creates a Microvm snapshot.
+pub fn create_snapshot(
+    vmm: &mut Vmm,
+    params: CreateSnapshotParams,
+) -> std::result::Result<(), CreateSnapshotError> {
+    let microvm_state = vmm
+        .save_state()
+        .map_err(CreateSnapshotError::MicrovmState)?;
+
+    snapshot_memory_to_file(vmm, &params.mem_file_path, params.snapshot_type)?;
+
+    snapshot_state_to_file(&microvm_state, &params.snapshot_path, params.version)?;
+
+    Ok(())
+}
+
+fn snapshot_state_to_file(
+    microvm_state: &MicrovmState,
+    snapshot_path: &PathBuf,
+    version: Option<u16>,
+) -> std::result::Result<(), CreateSnapshotError> {
+    let mut snapshot_file = OpenOptions::new()
+        .create(true)
+        .write(true)
+        .open(snapshot_path)
+        .map_err(CreateSnapshotError::SnapshotBackingFile)?;
+
+    // TODO: VersionMap should be managed at a higher level.
```

Comment:
We can manage it at the higher level, in that case define a parameter for it. 
The version map is not transient so It looks comfy to just use lazy static. @firecracker-microvm/compute-capsule WDYT ?

Reply:
That sounds fine

Reply:
:+1: